docs/architecture/adr-019-protobuf-state-encoding.md

```diff
@@ -110,162 +111,155 @@ type Codec interface {
+above, app's can still choose to use `oneof` to encode state and transactions
+but it is not the recommended approach.
+
+### Safe usage of `Any`
+
+By default, the [gogo protobuf implementation of `Any`](https://godoc.org/github.com/gogo/protobuf/types)
+uses global type registration to decode values packed in `Any` into concrete
+go types. This introduces a vulnerability where any malicious module
+in the dependency tree could registry a type with the global protobuf registry
+and cause it to be loaded and unmarshaled by a transaction that referenced
+it in the `type_url` field.
+
+To prevent this, we introduce a type registration mechanism for decoding `Any`
+values into concrete types through the `InterfaceContext` interface which
+bears some similarity to type registration with Amino:
 
-Example:
-
-```protobuf
-// app/codec/codec.proto
-
-import "third_party/proto/cosmos-proto/cosmos.proto";
-import "x/auth/types/types.proto";
-import "x/auth/vesting/types/types.proto";
-import "x/supply/types/types.proto";
-
-message Account {
-  option (cosmos_proto.interface_type) = "*github.com/cosmos/cosmos-sdk/x/auth/exported.Account";
-
-  // sum defines a list of all acceptable concrete Account implementations.
-  oneof sum {
-    cosmos_sdk.x.auth.v1.BaseAccount                      base_account               = 1;
-    cosmos_sdk.x.auth.vesting.v1.ContinuousVestingAccount continuous_vesting_account = 2;
-    cosmos_sdk.x.auth.vesting.v1.DelayedVestingAccount    delayed_vesting_account    = 3;
-    cosmos_sdk.x.auth.vesting.v1.PeriodicVestingAccount   periodic_vesting_account   = 4;
-    cosmos_sdk.x.supply.v1.ModuleAccount                  module_account             = 5;
-  }
-
-  // ...
+```go
+type InterfaceContext interface {
+    // RegisterInterface associates protoName as the public name for the
+    // interface passed in as iface
+    // Ex:
+	//   ctx.RegisterInterface("cosmos_sdk.Msg", (*sdk.Msg)(nil))
+	RegisterInterface(protoName string, iface interface{})
+
+    // RegisterImplementation registers impl as a concrete implementation of
+    // the interface iface
+    // Ex:
+	//   ctx.RegisterImplementation((*sdk.Msg)(nil), &MsgSend{})
+	RegisterImplementation(iface interface{}, impl proto.Message)
+
+    // UnpackAny unpacks the value in any to the interface pointer passed in as
+    // iface. Note that the type in any must have been registered with
+    // RegisterImplementation as a concrete type for that interface
+    // Ex:
+    //    var msg sdk.Msg
+    //    err := ctx.UnpackAny(any, &msg)
+    //    ...
+	UnpackAny(any *Any, iface interface{}) error
 }
 ```
 
-```go
-// app/codec/codec.go
-
-type Codec struct {
-  codec.Marshaler
+In addition to serving as a whitelist, `InterfaceContext` can also serve
+to communicate the list of concrete types that satisfy an interface to clients.
 
+Note that `InterfaceContext` usage does not deviate from standard protobuf
```

Comment:
How and where is this type used? It seems it's used in a concrete codec type. I'd note that here.

Reply:
Does this addition explain things clearly enough: https://github.com/cosmos/cosmos-sdk/pull/6081/commits/0470e4767a74fd3cccd7b10ba526252f6ca6cda3 ?